### PR TITLE
Fix MISRA C 2012 Rule 13.2 Violations

### PR DESCRIPTION
**Pull request by bradleysmith23**
Description
-----------
As stated in MISRA C 2012 - "Between any two sequence points or within any full expression (...) **There shall be no more than one read access with the volatile-qualified type**"
The recommended remediation by MISRA  is to only access volatiles with assignment statements whenever possible.

Test Steps
-----------

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tasks.c [hunk 1]
```diff
@@ -3188,7 +3188,13 @@ static void prvInitialiseNewTask( TaskFunction_t pxTaskCode,
...
                      * to by pxCurrentTCB has just been suspended and pxCurrentTCB
                      * must be adjusted to point to a different task. */
-                    if( listCURRENT_LIST_LENGTH( &xSuspendedTaskList ) == uxCurrentNumberOfTasks )
+
+                    /* Use temp variable as distinct sequence points for reading
+                     * volatile variables prior to a comparison, per the coding guidelines */
+                    UBaseType_t uxCurrentListLength = listCURRENT_LIST_LENGTH( &xSuspendedTaskList );
+                    UBaseType_t uxCurrentNumOfTasks = uxCurrentNumberOfTasks;
+
+                    if( uxCurrentListLength == uxCurrentNumOfTasks )
```
Comment (chinglee-iot) [inline tasks.c:3197]: Can we reduce one temp variable with the following example? Then only one volatile variable is used in comparison.
```suggestion
                    /* Use temp variable as distinct sequence points for reading
                     * volatile variables prior to a comparison, per the coding guidelines */
                    UBaseType_t uxCurrentListLength = listCURRENT_LIST_LENGTH( &xSuspendedTaskList );

                    if( uxCurrentListLength == uxCurrentNumberOfTasks )
```

Comment (bradleysmith23) [reply tasks.c:3197]: I do not have a strong preference towards either one, the reason I decided to assign both to temp variables was I thought it made the intent more clear.

**Review by chinglee-iot — APPROVED** [review]

**Review by chinglee-iot — APPROVED** [review]

File: tasks.c [hunk 2]
```diff
@@ -3188,7 +3188,12 @@ static void prvInitialiseNewTask( TaskFunction_t pxTaskCode,
                     /* The scheduler is not running, but the task that was pointed
                      * to by pxCurrentTCB has just been suspended and pxCurrentTCB
                      * must be adjusted to point to a different task. */
-                    if( listCURRENT_LIST_LENGTH( &xSuspendedTaskList ) == uxCurrentNumberOfTasks )
+
+                    /* Use temp variable as distinct sequence points for reading
+                     * volatile variables prior to a comparison, per the coding guidelines */
+                    UBaseType_t uxCurrentListLength = listCURRENT_LIST_LENGTH( &xSuspendedTaskList );
```
Comment (aggarg) [inline tasks.c:3194]: For consistency, we should declare the variable at the beginning of the function.

Comment (aggarg) [inline tasks.c:3194]:
```suggestion

                    /* Use a temp variable as distinct sequence points for reading
                     * volatile variables prior to a comparison to ensure compliance
                     * with MISRA C 2012 Rule 13.2. */
                    UBaseType_t uxCurrentListLength = listCURRENT_LIST_LENGTH( &xSuspendedTaskList );
```

File: tasks.c [hunk 3]
```diff
@@ -3236,7 +3238,13 @@ static void prvInitialiseNewTask( TaskFunction_t pxTaskCode,
                     /* The scheduler is not running, but the task that was pointed
                      * to by pxCurrentTCB has just been suspended and pxCurrentTCB
                      * must be adjusted to point to a different task. */
-                    if( listCURRENT_LIST_LENGTH( &xSuspendedTaskList ) == uxCurrentNumberOfTasks )
+
+                    /* Use temp variable as distinct sequence points for reading
```
Comment (kar-rahul-aws) [inline tasks.c:3242]: NIT
/* Use **a** temp variable as **a** distinct sequence point */

**Review by kar-rahul-aws — APPROVED** [review]

**Review by aggarg — APPROVED** [review]

**Pull request merged**